File: data-validation/scala/quick-start/metrics/src/main/scala/com/here/platform/data/validation/example/quickstart/metrics/scala/Calculator.scala
/*
 * Copyright (C) 2017-2020 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.platform.data.validation.example.quickstart.metrics.scala

import com.here.platform.data.processing.driver.Default
import com.here.platform.data.processing.catalog.Partition.Generic
import com.here.platform.data.processing.compiler.{InKey, OutKey}
import com.here.platform.data.validation.example.quickstart.common.LayerDefs
import com.here.platform.schema.data.validation.core.metrics.v1.metrics.TestMetricSeverity
import com.here.platform.schema.data.validation.core.testing.v1.testing.TestResult.TestResultType
import com.here.platform.schema.data.validation.example.quickstart.metrics.v1.metrics.{
  Metric,
  MetricSample
}

// The actual map-reduce implementation that assigns severities to the test results
class Calculator(val metricsConfig: MetricsConfig)
    extends com.here.platform.data.validation.core.metrics.calculators.Metric[Data, Metric] {
  val id: String = "quickstartmetriccalc"

  /** Maps input data into output partitions, restructuring into new feature contexts as needed.
    *
    * @param context    Feature context for the given input partition
    * @return Feature contexts for each output partition mapped from the input data
    */
  override def map(context: (InKey, Data)): Iterable[(OutKey, Data)] = {
    val outKey = OutKey(catalog = Default.OutCatalogId,
                        layer = LayerDefs.Metrics.Out.metricsResultLayerName,
                        partition = Generic(context.value.result.result.value.toString))
    Iterable((outKey, context.value))
  }

  /** Receives feature contexts from mapper and packages them into output protobufs.
    *
    * @param data       All the feature contexts for the given output partition, as produced by the mapper
    * @return The reduction output as optional
    */
  override def reduce(data: (OutKey, Iterable[Data])): Option[Metric] = {
    val tileIds = data.value.map(_.tileID).toSeq
    val isPass = data.key.partition.asInstanceOf[Generic].name.equals(TestResultType.PASS.name)
    if (tileIds.nonEmpty) {
      if (isPass && !metricsConfig.writeSeverityNoneSamples) {
        // If this is a PASS partition and we are not configured to write SeverityNone samples, we can skip writing output.
        None
      } else {
        // PASS results are considered SeverityNone; all other statuses are CRITICAL
        val metricSample: MetricSample = MetricSample(
          severity = TestMetricSeverity(metricid = id,
                                        value =
                                          if (isPass) TestMetricSeverity.SeverityType.NONE
                                          else TestMetricSeverity.SeverityType.CRITICAL),
          tileIds = tileIds
        )
        Some(Metric(metrics = Seq(metricSample)))
      }
    } else {
      // If we encountered no tile IDs, we have nothing to output.
      None
    }
  }
}

File: data-validation/java/quick-start/metrics/src/main/java/com/here/platform/data/validation/example/quickstart/metrics/java/Calculator.java
/*
 * Copyright (C) 2017-2021 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.platform.data.validation.example.quickstart.metrics.java;

import com.here.platform.data.processing.driver.Default;
import com.here.platform.data.processing.java.Pair;
import com.here.platform.data.processing.java.catalog.partition.Generic;
import com.here.platform.data.processing.java.catalog.partition.Key;
import com.here.platform.pipeline.logging.LogContext;
import com.here.platform.schema.data.validation.core.metrics.v1.Metrics.TestMetricSeverity;
import com.here.platform.schema.data.validation.core.testing.v1.Testing.TestResult.TestResultType;
import com.here.platform.schema.data.validation.example.quickstart.metrics.v1.Metrics.Metric;
import com.here.platform.schema.data.validation.example.quickstart.metrics.v1.Metrics.MetricSample;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

// The actual map-reduce implementation that assigns severities to the test results
public class Calculator
    implements com.here.platform.data.validation.core.java.metrics.calculators.Metric<
        Data, Metric> {

  private static final long serialVersionUID = -4545526780955027307L;

  private MetricsConfig config;

  public Calculator(MetricsConfig metricsConfig) {
    config = metricsConfig;
  }

  @Override
  public String id() {
    return "quickstartmetriccalc";
  }

  @Override
  public Iterable<Pair<Key, Data>> map(Pair<Key, Data> context, LogContext logContext) {
    // The input test result is mapped to an output partition with the name of the result.
    // In this example, all FAIL results from the input catalog go into the FAIL partition.
    Generic resultAsString =
        new Generic(context.getValue().getResult().getResult().getValue().toString());
    Key outKey = new Key(Default.OutCatalogId().name(), "metrics-result", resultAsString);
    return Collections.singletonList(new Pair<>(outKey, context.getValue()));
  }

  @Override
  public java.util.Optional<Metric> reduce(Pair<Key, Iterable<Data>> data, LogContext logContext) {

    // Aggregate all tile IDs from this type of test result. For this example, the type is always
    // FAIL.
    List<Long> tileIds =
        StreamSupport.stream(data.getValue().spliterator(), true)
            .map(Data::getTileID)
            .sorted()
            .collect(Collectors.toList());

    if (!tileIds.isEmpty()) {
      TestMetricSeverity.SeverityType severityType = TestMetricSeverity.SeverityType.CRITICAL;

      if (data.getKey().partition().toString().equals(TestResultType.PASS.toString())) {
        // PASS results are considered Severity NONE
        if (config.getWriteSeverityNoneSamples())
          severityType = TestMetricSeverity.SeverityType.NONE;
        else return java.util.Optional.empty();
      }

      // In this example, the calculator assigns a CRITICAL severity for all non-PASS results,
      // and collects all the tile IDs for each result type in the input catalog.
      MetricSample metricSample =
          MetricSample.newBuilder()
              .addAllTileIds(tileIds)
              .setSeverity(
                  TestMetricSeverity.newBuilder().setMetricid(id()).setValue(severityType).build())
              .build();
      return java.util.Optional.of(
          Metric.newBuilder().addAllMetrics(Collections.singletonList(metricSample)).build());
    }

    // If no tile IDs were encountered, there is nothing to output
    return java.util.Optional.empty();
  }
}
